scripts/analyze_memleaks.patch

```diff
@@ -47,18 +47,19 @@ index c73c24cf..fbc704f7 100644
 -void Memory_FreePointer(void *arg)
 +void Memory_FreePointerImpl(void *arg, const char *file, int line, const char *func)
```

Comment:
sorry my 20hrs a week job has got a horror project that turned into a 40 + weekends. So I've not been able to look through all of FFMPEG lib code to see if they have also made this error. After OPEN_SSL anything is possible. 

This issue is made worse by the fact that gcc is a stickler for one of C's golden rules, and that is 'anything can be converted to a void* no questions asked', so it won't give you or anybody a warning about this. If you put it through g++ it will complain heavily, I'm told. These should be `void** arg`. As you are calling them with `Memory_FreePointer(&pointer);` which is correct, just the `void**` is getting converted to a `void*`.  Thus when you do `*pointer` via the `memcopy` you get a `void*` and not a `void` but the type is actually `void`. Memcpy is not type safe in anyway was so happily takes your `void` and reads `void*` number of bytes anyway because you told it that many bytes and you just so happened to pass in a `void**` so it was ok. Sadly gcc will not give you any warning about forgetting the & and doing `Memory_FreePointer(MyPointer)` will compile just fine. Even if the function is a `void**` and you pass a `void*` to it ; still no error.

From what I've looked at in FFMPEG they typically have REF forms of their objects so `AudioBufferREF` where in that is a `typedef` to `AudioBuffer*` and hence when they do `Free(AudioBufferREF* pointer)` that is actually a `Free(AudioBuffer** pointer)` which does give you a warning. For example 
```
#include <stdlib.h>
#include <string.h>

struct tempStruct {
    int i;
};

typedef struct tempStruct* MyRef;

void Memory_Free(void *memory)
{
    if (memory) {
        free(memory);
    }
}

void Memory_FreePointer(MyRef* arg)
{
    void *memory;
    memcpy(&memory, arg, sizeof(void *));
    memcpy(arg, &(void *) { NULL }, sizeof(void *));
    Memory_Free(memory);
}

void main()
{
    struct tempStruct temp = { 4 };
    MyRef pTS = &temp;
    Memory_FreePointer(pTS);
}
```
will give you 
```
<source>(29): warning C4047: 'function': 'MyRef *' differs in levels of indirection from 'MyRef'
<source>(29): warning C4024: 'Memory_FreePointer': different types for formal and actual parameter 1
````
from MSVC and
```
<source>:29:24: warning: incompatible pointer types passing 'MyRef' (aka 'struct tempStruct *') to parameter of type 'MyRef *' (aka 'struct tempStruct **'); take the address with & [-Wincompatible-pointer-types]
    Memory_FreePointer(pTS);
                       ^~~
                       &
<source>:17:32: note: passing argument to parameter 'arg' here
void Memory_FreePointer(MyRef* arg)
```
from Clang and 
```
<source>:29:24: warning: passing argument 1 of 'Memory_FreePointer' from incompatible pointer type [-Wincompatible-pointer-types]
   29 |     Memory_FreePointer(pTS);
      |                        ^~~
      |                        |
      |                        MyRef {aka struct tempStruct *}
<source>:17:32: note: expected 'struct tempStruct **' but argument is of type 'MyRef' {aka 'struct tempStruct *'}
   17 | void Memory_FreePointer(MyRef* arg)
      |                         ~~~~~~~^~~
```
from GCC. Making it much safer.  I don't think the REF typedef is really suitable for this case though, but at least having the ** so people can see that its meant to be an & if they fall victim to it, or a MACRO that applies the & for them would be a more suitable solution for this code base.

Reply:
I agree that it leaves an easy opportunity to shoot oneself in the foot but I perceive the majority of C to be like this. The signature was chosen to use `void*` precisely to use C's implicit casting and avoid warnings such as:

```
char *string = NULL;
Memory_FreePointer(&string);
```

```
/home/rr-/test.c: In function ‘main’:
/home/rr-/test.c:27:24: warning: passing argument 1 of ‘Memory_FreePointer’ from incompatible pointer type [-Wincompatible-pointer-types]
   27 |     Memory_FreePointer(&string);
      |                        ^~~~~~~
      |                        |
      |                        char **
/home/rr-/test.c:16:32: note: expected ‘void **’ but argument is of type ‘char **’
   16 | void Memory_FreePointer(void **arg)
```

whereas it's the intended use of this function. I'd prefer to stay away from macros as much as possible (compare: `Memory_FreePointer(&ptr)` vs `MEMORY_FREE_PTR(ptr);`). Do we have any alternative?

Reply:
I've been thinking about alternatives but then I remembered 
```
#define Memory_FreePointer(memory)      (Memory_FreePointerImpl(memory, __FILE__, __LINE__, __func__))
```
you already have a macro, that is not all uppercase, because macros don't need to be upper case by way of C.  
However an annoying problem is gcc is fine with 
```
void*  to void** fine
void** to void*  fine
char** to void*  fine
char** to void** not fine
```
is a bit of an issue, because the macro needs to become
```
#define Memory_FreePointer(memory)      (Memory_FreePointerImpl((void**)&memory, __FILE__, __LINE__, __func__))
```
which throws away all type safety as we cast. However having the functions porotype and definition have ** so if anybody looks it, they will be able understand how it is used, rather than looking at it, seeing they are doing the right thing and wondering why they have a really strange crash with it, I feel is the lesser of the two evils. 
But then also at this point you could just avoid the complicated setup, performance hit of doing a few mem copies and just 
```
#define Memory_FreePointer(memory)     {Memory_FreePointerImpl(memory, __FILE__, __LINE__, __func__); memory=NULL;}
```
and the function just takes a `void*`

Reply:
This macro actually only exists in a .patch that's supposed to help trace memory leaks; normally we use `UPPER_CASE`. The broken name convention is to make the patch minimal in size.

Unfortunately, none of the alternatives look appealing to me. The proposals that add taking a reference underneath actually would confuse me more than the current approach. I'll add some documentation on this function and its intended usage.